File: LeetCode/86.partition-list.cpp
/*
 * @lc app=leetcode id=86 lang=cpp
 *
 * [86] Partition List
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
public:
    ListNode *partition(ListNode *head, int x)
    {
        ListNode dummy(0);
        ListNode *cur = head, *tail1 = &dummy, *tail2 = nullptr;
        while (cur != nullptr)
        {
            if (cur->val >= x)
            {
                if (tail2 != nullptr)
                {
                    tail2->next = cur;
                }
                else
                {
                    tail1->next = cur;
                }
                tail2 = cur;
                cur = cur->next;
                tail2->next = nullptr;
            }
            else
            {
                ListNode *next = cur->next;
                cur->next = tail1->next;
                tail1->next = cur;
                tail1 = cur;
                cur = next;
            }
        }
        return dummy.next;
    }
};

class Solution
{
public:
    ListNode *partition(ListNode *head, int x)
    {
        ListNode less(0), gt(0);
        ListNode *lessPtr = &less, *gtPtr = &gt;
        while (head != nullptr)
        {
            if (head->val < x)
            {
                lessPtr = lessPtr->next = head;
            }
            else
            {
                gtPtr = gtPtr->next = head;
            }
            head=head->next;
        }
        gtPtr->next = nullptr;
        lessPtr->next = gt.next;
        return less.next;
    }
};

File: LeetCode/java/86.partition-list.java
/*
 * @lc app=leetcode id=86 lang=java
 *
 * [86] Partition List
 *
 * https://leetcode.com/problems/partition-list/description/
 *
 * algorithms
 * Medium (41.08%)
 * Likes:    1772
 * Dislikes: 354
 * Total Accepted:    243.8K
 * Total Submissions: 566.7K
 * Testcase Example:  '[1,4,3,2,5,2]\n3'
 *
 * Given the head of a linked list and a value x, partition it such that all
 * nodes less than x come before nodes greater than or equal to x.
 * 
 * You should preserve the original relative order of the nodes in each of the
 * two partitions.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: head = [1,4,3,2,5,2], x = 3
 * Output: [1,2,2,4,3,5]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: head = [2,1], x = 2
 * Output: [1,2]
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [0, 200].
 * -100 <= Node.val <= 100
 * -200 <= x <= 200
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null)
            return null;
        ListNode dummyLt = new ListNode();
        ListNode dummyGt = new ListNode();
        ListNode lt = dummyLt, gt = dummyGt;

        while (head != null) {
            ListNode rest = head.next;
            if (head.val < x) {
                lt.next = head;
                lt = lt.next;
            } else {
                gt.next = head;
                gt = gt.next;
            }
            head = rest;
        }
        lt.next = dummyGt.next;
        gt.next = null;
        return dummyLt.next;
    }
}
// @lc code=end
